iris/src/iris/pipeline/shared/citation_pipeline.py

```
@@ -27,32 +25,165 @@ class InformationType(str, Enum):
+
+def _extract_marker_indices(answer_with_markers: str) -> set[int]:
+    return {int(m.group(1)) for m in _MARKER_RE.finditer(answer_with_markers)}
+
+
+def _validate_marker_coverage(resp: CitationPromptResponse) -> None:
+    marker_indices = _extract_marker_indices(resp.answer_with_markers)
+    citation_indices = [c.index for c in resp.citations]
+    cited_set = set(citation_indices)
+
+    if len(citation_indices) != len(cited_set):
+        raise ValueError("Duplicate citation indices in citations array")
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)